### ISSUE TITLE
Suggestion: replace `MessageCodec` named methods for Runtime with single generic method 

### ISSUE THREAD

**Issue by citizen-stig** [Sovereign-Labs/sovereign-sdk#389]
# Problem

Here's is simple runtime with 3 modules:

```rust
use sov_modules_api::Context;
use sov_modules_macros::{DispatchCall, Genesis, MessageCodec};


#[derive(Genesis, DispatchCall, MessageCodec)]
#[serialization(borsh::BorshDeserialize, borsh::BorshSerialize)]
pub struct Runtime<C: Context> {
    pub accounts: sov_accounts::Accounts<C>,

    pub bank: sov_bank::Bank<C>,

    pub value_setter: sov_value_setter::ValueSetter<C>,
}
```

If client wants to submit call message, she needs to encode it for this runtime. Each module has its own "encode" method:

* `encode_accounts_call(msg: accounts::call:CallMessage) -> Vec<u8>;`
* `encode_bank_call(msg: accounts::call:CallMessage) -> Vec<u8>;`
* `encode_value_setter_call(msg: accounts::call:CallMessage) -> Vec<u8>;`

And then encoded vector needs to be signed.

It might be cumberstone to have specific method names for each method, and makes code generation for client a little more compilcated.

# Suggestion

As `encode_` methods are generated by macro, why not make it generic, by generating trait and implmenting it for each call message:

```rust

trait<T> EncodeCallMessage {
    fn encode(call_mesasge: T) -> Vec<u8>;
};

impl<accounts::call:CallMessage, C: Context> EncodeCallMessage for  Runtime<C> {
    fn encode(call_mesasge: T) -> Vec<u8> {
        // actual implementation, similar to what is done now 
    }
}
```

**Comment by citizen-stig** [comment]
@bkolad @preston-evans98 @dubbelosix , this is suggestion, let's discuss it over daily.

**Comment by citizen-stig** [comment]
@bkolad Could you please elaborate, what adding associated context type will give ?

**Comment by preston-evans98** [comment]
Nice suggestion @citizen-stig! 

Do we need the trait to be generic at all? Seems like we can just get away with: 


```rust
trait EncodeCallMessage {
    type CallMessage;
    fn encode(call_message: &CallMessage) -> Vec<u8>;
};
```

cc @@bkolad

**Comment by bkolad** [comment]
We need to generate multiple implementations for `CallMessage` from each module defined in the `Runtime`.

```Rust
impl EncodeCallMessage for Runtime{
  type CallMessage = bank::CallMessage;
   ...
}

impl EncodeCallMessage for Runtime{
  type CallMessage = accounts::CallMessage;
  ...
}
```
won't compile since the two impl differ only by an associate type (and rust won't allow it)
This will compile:

```Rust
impl EncodeCallMessage<bank::CallMessage;> for Runtime{
 ...
}

impl EncodeCallMessage<accounts::CallMessage> for Runtime {
  ...
}
```
We never implement `EncodeCallMessage` for two different `Contexts` at the same time (the `Context` is just taken from runtime) so the above limitation doesn't apply to `Context` therefore we should make it an associated type (which is the preferred way because it helps with type inference)

**Comment by preston-evans98** [comment]
Oh yeah, of course! I was thinking that the trait would go on the module - which at second glance makes no sense. The generic trait is definitely the way to go here.

**Comment by citizen-stig** [comment]
Assigning to double check later that it was impleneted in #598

**Comment by citizen-stig** [comment]
Confirmed that it is implemented as part of #598
